### ISSUE TITLE
Bad Request when using POST [2.x]

### ISSUE THREAD

**Issue by alexgkt** [RESTful-Drupal/restful#762]
Created a custom entity and have no issue to GET from the endpoint but having issue on a POST. Kept getting this message. Anything else i've missed?

{"type":"http:\/\/www.w3.org\/Protocols\/rfc2616\/rfc2616-sec10.html#sec10.4.1","title":"Property name cannot be set.","status":400,"detail":"Bad Request"}

Here's how my plugin is being defined

<pre lang="php"><code>
namespace Drupal\tuners\Plugin\resource\model\v1;

use Drupal\restful\Plugin\resource\ResourceInterface;
use Drupal\restful\Plugin\resource\ResourceEntity;

/**
 * Class Models
 * @package Drupal\restful\Plugin\resource
 *
 * @Resource(
 *   name = "models:1.0",
 *   resource = "models",
 *   label = "Models",
 *   description = "Export the model entity type.",
 *   authenticationTypes = TRUE,
 *   authenticationOptional = TRUE,
 *   dataProvider = {
 *     "entityType": "model",
 *     "idField": "id"
 *   },
 *   majorVersion = 1,
 *   minorVersion = 0
 * )
 */
class Models__1_0 extends ResourceEntity implements ResourceInterface {

  protected function publicFields() {
    $fields = parent::publicFields();

    $fields['name'] = array(
      'property' => 'name',
    );

    $fields['trim'] = array(
      'property' => 'trim',
    );

    $fields['make'] = array(
      'property' => 'make',
    );

    $fields['year_start'] = array(
      'property' => 'year_start',
    );

    $fields['year_end'] = array(
      'property' => 'year_end',
    );

    return $fields;
  }

}</code></pre>

all $fields are property items defined on hook_entity_info

**Comment by e0ipso** [comment]
Can you share the post payload you are sending?

**Comment by alexgkt** [comment]
Here's the payload

{id: "6", name: "1", trim: "1", make: "Subaru", year_start: "2001", year_end: null, year_start: "2001"}

**Comment by e0ipso** [comment]
Yep, posting a record with an ID is a bad request. Try dropping the ID or using PUT.

Let me know how it goes.

**Comment by alexgkt** [comment]
Same result with this payload: 

``` json
{"name":"1","make":"1","trim":"1","year_start":"1","year_stop":"1","changed":"2015-12-02 23:39:34","created":"2015-12-02 23:39:34"}
```

Getting this as response:

``` json
{"type":"http:\/\/www.w3.org\/Protocols\/rfc2616\/rfc2616-sec10.html#sec10.4.1","title":"Property name cannot be set.","status":400,"detail":"Bad Request"}
```

When i use PUT i got this instead:

``` json
{"type":"http:\/\/www.w3.org\/Protocols\/rfc2616\/rfc2616-sec10.html#sec10.4.1","title":"The http method PUT is not allowed for this path.","status":400,"detail":"Bad Request"}
```

**Comment by alexgkt** [comment]
Not sure if its helps but here's the header:

POST /api/v1.0/models HTTP/1.1
Host: localhost:8002
Connection: keep-alive
Content-Length: 131
Accept: application/json, text/plain, _/_
Origin: http://localhost:8002
X-CSRF-Token: REfHxFm-PaoTphaSh7YpsUFOzOwkXQjw5DWnKKS5MQo
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36
Content-Type: application/json;charset=UTF-8
Referer: http://localhost:8002/admin/tuners/models?u=/admin/tuners/models
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.8
Cookie: FirstTimer=1; PHPSESSID=kstain6cjj7t5agpp4as883sv1; SESS14ceab87d7c920bcb7fb65c2d2176de3=wrD5Pdo4AeRLaoHEWGophKMZotNQdgzOZ-VofEjSb60; Drupal.tableDrag.showWeight=0; __atuvc=6%7C44%2C0%7C45%2C0%7C46%2C94%7C47%2C25%7C48; SESSbaba9105a828fc1d19f821cdd448e29e=xmGF68UssJnhDLh3dnfkbt9Kvc_FdSRwm7jn_s5CMko; has_js=1

**Comment by e0ipso** [comment]
`"title":"Property name cannot be set."` <--- You did not send a title.

> When i use PUT i got this instead:

You need to use PUT in the correct path instead.

**Comment by alexgkt** [comment]
But the custom entity i defined doesn't have a title column. 

<pre lang="php"><code>
$info['model'] = array(
    'label' => t('Model'),
    'base table' => 'car_model',
    'entity keys' => array(
      'id' => 'id',
      'label' => 'name',
    ),
    'module' => 'tuners',
    'entity class' => 'Entity',
    'controller class' => 'EntityAPIController',
    'access callback' => 'tuners_model_access_callback',
    'uri callback' => 'entity_class_uri',
    'views controller class' => 'EntityDefaultViewsController',
    //'file' => 'tuners_entities.inc',
  );
</code></pre>

Anyway try to add a name property to the payload but still no go:
{"name":"1","make":"1","trim":"1","year_start":"1","year_stop":"1","changed":"2015-12-02 23:59:26","title":"test","created":"2015-12-02 23:59:26"}

Same error

**Comment by laradevitt** [comment]
This situation is nearly identical to mine, and it was at least partially resolved by a [pull request](https://github.com/RESTful-Drupal/restful/pull/754) referenced in #744.  

See [my comment](https://github.com/RESTful-Drupal/restful/issues/744#issuecomment-248033028) for details.

**Comment by laradevitt** [comment]
OK, never mind that last comment. It will likely allow you to create an entity, but the field values will probably be empty. 

(@alexgkt) If you're basing your entity type on the [Model Entities](https://www.drupal.org/project/model) module, you will need to explicitly declare a 'setter callback' for your fields. You can do that by [implementing `hook_entity_property_info()](http://drupal.stackexchange.com/a/134453/28700) or  [overriding EntityDefaultMetadataController](https://www.drupal.org/node/2220557).
